File: C++_Code_Solutions/Reverse Linked List.cpp
/* Reverse a singly linked list.

Example:

Input: 1->2->3->4->5->NULL
Output: 5->4->3->2->1->NULL
Follow up:

A linked list can be reversed either iteratively or recursively. Could you implement both?
 */

// Done by >> Anand Kothari

/*
  Definition for singly-linked list.
  struct ListNode {
      int val;
      ListNode *next;
      ListNode(int x) : val(x), next(NULL) {}
  };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {

        // RECURSION METHOD

        //  if (head==NULL || head->next==NULL) return head;
        // ListNode *h = reverseList(head->next);
        // head->next->next = head;
        // head->next = NULL;
        // return h;


        // Iterative method
        ListNode *prev = NULL , *current = head, *next = NULL;

        while(current != NULL)
        {
           // Store next
            next = current->next;

            // Reverse current node's pointer
            current->next = prev;

            // Move pointers one position ahead.
            prev = current;
            current = next;
        }
        head = prev ;
        return head;
    }
};

File: Python_Code_Solutions/Reverse_Linked_List.py

'''
Done By > Anand Kothari

Question::
Reverse a singly linked list.

Example:

Input: 1->2->3->4->5->NULL
Output: 5->4->3->2->1->NULL

--------------------------------------
Solution: 

'''
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# Simple iteratively without extra space

# class Solution(object):
#     def reverseList(self, head):
#         prev = None 
#         curr = head
#         while curr!=None :
#             temp = curr.next
#             curr.next = prev
#             prev = curr
#             curr = temp
#         head = prev
#         return (head)
        
# Recursively

class Solution:
    def reverseList(self, head):
        
        if head is None or head.next is None:
            return head
        
        p = self.reverseList(head.next)
        head.next.next= head
        head.next = None        # It removes the unwanted pointer 
        return p
